numpyro/infer/inspect.py

```diff
@@ -364,16 +366,24 @@ def get_log_probs(sample):
             data=sample
```

Comment:
Thanks @YanniPapandreou for the clear explanation! It was my mistake - suggested updating `substitute` handler. You can revert that change and use the follow handler instead.
```python
class substitute_deterministic(handlers.substitute):
    def process_message(self, msg):
        if msg["type"] == "deterministic":
            msg["args"] = (msg["value"],)
            msg["value"] = self.data.get(msg["name"])

with handlers.trace() as tr, handlers.seed(rng_seed=0):
    with handlers.substitute(data=sample), substitute_deterministic(data=sample):
        model(*model_args, **model_kwargs)
```
Essentially, we are moving the actual value (obtained from the computation over other values) to `msg["args"]`, then in `provenance_arrays` below, we can do
```
            elif site["type"] == "deterministic":
                provenance_arrays[name] = site["args"][0]
```
This way we can know which variables that this deterministic variable depends on.

Reply:
> Thanks @YanniPapandreou for the clear explanation! It was my mistake - suggested updating `substitute` handler. You can revert that change and use the follow handler instead.

I have reverted this, for reference please see [here](https://github.com/pyro-ppl/numpyro/pull/1508/commits/e08a561c43f2a1fce684845a009faeb634ee05cd).

I have also created the handler you suggested in the `get_log_probs` function [here](https://github.com/pyro-ppl/numpyro/pull/1508/commits/ce10130630e3d6a8ef32c294d3bd209fa0cb88b3). I was getting errors with the code as is so I added the following key,value pairs to the message when it is a deterministic site:

```python
class substitute_deterministic(handlers.substitute):
            def process_message(self, msg):
                if msg["type"] == "deterministic":
                    msg["args"] = (msg["value"],)
                    msg["kwargs"] = {}
                    msg["value"] = self.data.get(msg["name"])
                    msg["fn"] = lambda x: x
```
This seems to work on the example.